server/src/main/java/org/elasticsearch/index/mapper/DynamicTemplate.java

```diff
@@ -257,7 +260,8 @@ static DynamicTemplate parse(String name, Map<String, Object> conf) throws Mappe
         List<String> pathUnmatch = new ArrayList<>(4);
         Map<String, Object> mapping = null;
         boolean runtime = false;
-        String matchMappingType = null;
+        List<String> matchMappingType = new ArrayList<>(4);
```

Comment:
Why is the size of the array 4? Just to have a size that is lower than he default?

Reply:
This uses the same initial size of the array list as the other lists a few lines above. I think this is because we expect most of these to have one or a few values, so a custom default value 4 aims for more efficiency than the generic default value.